### ISSUE TITLE
spatSample of large and sparse rasters

### ISSUE THREAD

**Issue by PetiteTong** [rspatial/terra#905]
Can you implement a separate method for `spatSample` of large and sparse rasters?
See: https://github.com/rspatial/terra/issues/557#issuecomment-1324452073

**Comment by rhijmans** [comment]
I have added a new argument, `exhaustive=FALSE`, to `spatSample`. Setting the argument to `TRUE` should allow you to deal with this issue. Can you try?

```
a1 = terra::spatSample(x1, 2000, as.df=T, replace=T, na.rm=T, xy=T, exhaustive=TRUE)
```

You need terra 1.6-45. You can install it with 

```
install.packages('terra', repos='https://rspatial.r-universe.dev')
```

**Comment by PetiteTong** [comment]
Dear rhijmans, thanks a lot!
 I tried terra 1.6-45,  a1 = terra::spatSample(x1, 2000, as.df=T, replace=T, na.rm=T, xy=T, exhaustive=TRUE)
 it works. Thank you for your wisdom and hard working!

Best, 
Petite Tong

**Comment by ManuelPopp** [comment]
I still have this issue:

When I check how many pixels I have per stratum, I get a large enough number:
`
> freq(boreal_union_classified)
  layer value     count
1     1     1   8906148
2     1     2 144431094
`

However, when I then try to draw a sample:
`
n_samples <- 5000
samples <- terra::spatSample(
  boreal_union_classified, size = n_samples, method = "stratified",
  replace = FALSE, na.rm = TRUE, as.points = TRUE, values = TRUE,
  exhaustive = TRUE
)
`
I get a warning and if I count the number of points in the resulting `SpatVector`, it is smaller than expected. For the smaller of the two classes, I only get 1339 instead of 5000 values.
`
Warning messages:
1: [is.lonlat] coordinates are out of range for lon/lat 
2: [is.lonlat] coordinates are out of range for lon/lat 
3: [spatSample] fewer samples than requested for group(s): 1

> length(which(samples$CHELSA_kg0_1951.1980_V.2.1 == 2))
[1] 5000
> length(which(samples$CHELSA_kg0_1951.1980_V.2.1 == 1))
[1] 1339
`

I tried with `exhaustive = TRUE` as well as the default `exhaustive = FALSE`. The raster I am using contains mostly `NA` values. None of the settings gave me the requested number of samples...

For me, I wrote this work-around:
`
sample_list <- list()
for (class in terra::unique(boreal_union_classified)[, 1]) {
  idx <- which(terra::values(boreal_union_classified, mat = FALSE) == class)
  ilocs <- sample(idx, size = n_samples)
  sample_list[[class]] <- terra::vect(
    terra::xyFromCell(boreal_union_classified, cell = ilocs),
    crs = "epsg:4326"
  )
}

sample <- terra::vect(do.call(c, sample_list))
`

but this is horribly slow and I feel like the `SpatSample` function should somehow handle this...

Am I doing something wrong, do I expect the function to do something it is not supposed to do, or is there a bug?
